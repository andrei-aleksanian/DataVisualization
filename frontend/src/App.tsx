import Visualization2D from './Visualization2D';

// Returning this for now. Expecting to grow the application to return a router.
const App = () => <Visualization2D />;

export default App;
