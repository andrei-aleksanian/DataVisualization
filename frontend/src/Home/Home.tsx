import { LinkHero } from 'Components/Link';
import classes from './Home.module.scss';

export const TEXT_LINK_LIBRARY = 'See an Example';
export const TEXT_LINK_CUSTOM_DATA = 'Visualize Your Data';

export const TEXT_H1 = 'COVA/ANGEL Visualization';
export const TEXT_P =
  "Dimensionality reduction for cohort data visualization is to plot a high-dimensional dataset in 2D,3D spaces to help analysts identify data properties and gain insights visually. COVA and ANGEL are two novel data visualization algorithms that provide a simple but effective way to improve cohort data visualization performance comprehensively. This demo website illustrates examples embedded by these two algorithms under different hyper-parameter settings and supports users in visualizing their datasets. In addition, the fast-ANGEL version is applied to the users' custom datasets. The current restriction of input data format is '.mat'.";
const Home = () => {
  return (
    <div className={classes.index}>
      <div className={classes.Container}>
        <h1>{TEXT_H1}</h1>
        <p>{TEXT_P}</p>
        <div className={classes.LinkWrapper}>
          <LinkHero text={TEXT_LINK_LIBRARY} link="/examples" />
          <LinkHero text={TEXT_LINK_CUSTOM_DATA} link="/your-data" />
        </div>
      </div>
    </div>
  );
};

export default Home;
