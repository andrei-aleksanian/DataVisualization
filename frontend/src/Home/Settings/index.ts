export { default } from './Settings';
export * from './Settings';
export * from './components/ANGEL';
export * from './components/COVA';
