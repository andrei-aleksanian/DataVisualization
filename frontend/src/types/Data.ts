export type Points2D = [number, number, 0];

export type Points3D = [number, number, number];

// Data IN, uncolored:
export interface Data {
  labels: number[];
}

export interface Data2D extends Data {
  points: Points2D[];
}

export interface Data3D extends Data {
  points: Points3D[];
}

// Data here is not used atm
export const DATA: Data2D = {
  points: [
    [0, 0, 0],
    [0, 1, 0],
    [0, 2, 0],
    [3, 1, 0],
    [3, 2, 0],
    [3, 1, 0],
    [-3, 4, 0],
    [-3, 1, 0],
  ],
  labels: [1, 1, 1, 0, 0, 0, 3, 3],
};

// Data WITHIN React, colored:

export interface DataColored {
  colors: string[];
}

export interface Data2DColored extends DataColored {
  points: Points2D[];
}
export interface Data3DColored extends DataColored {
  points: Points3D[];
}
