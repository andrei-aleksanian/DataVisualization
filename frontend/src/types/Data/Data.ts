export type Point2D = [number, number, 0];
export type Point3D = [number, number, number];
export type Points = Point2D[] | Point3D[];

// Data common properties

export interface DataCore {
  dimension2D: boolean;
  iteration: number;
  maxIteration: number;
  g: number[][];
  labels: number[];
  Relation: number[][];
  Ad: number[][];
  V: number[][];
  points: Points;
  alpha: number;
}

// Data IN, with labels and without colors:
export interface DataLabelled extends DataCore {}

// dummy example data
export const DATA: DataLabelled = {
  dimension2D: true,
  // todo: double check
  points: [
    [0, 0, 0],
    [0, 1, 0],
    [0, 2, 0],
    [3, 1, 0],
    [3, 2, 0],
    [3, 1, 0],
    [-3, 4, 0],
    [-3, 1, 0],
  ],
  g: [[]],
  labels: [1, 1, 1, 0, 0, 0, 3, 3],
  Relation: [[]],
  Ad: [[]],
  V: [[]],
  alpha: 0.5,
  iteration: 0,
  maxIteration: 0,
};

// Data colored without labels
export interface DataColored extends DataLabelled {
  colors: string[];
}
