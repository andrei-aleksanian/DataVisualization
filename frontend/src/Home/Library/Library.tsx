import { useEffect, useState } from 'react';
import { LinkBack } from 'Components/Link';
import { ExampleProps } from 'types/Examples';

import { useHistory } from 'react-router-dom';
import getId from 'utils/getId';
import { Popup } from 'Components/UI';
import fetchExamples from './services';

import classes from './Library.module.scss';

export const TEXT_H1 = 'Examples Library';
export const TEXT_POPUP = "Couldn't load examples, please try again later";

const Library = () => {
  const [examples, setExamples] = useState([] as ExampleProps[]);
  const [error, setError] = useState(null as null | string);
  const history = useHistory();

  const Example = ({ name, description, id, imagePath }: ExampleProps) => (
    <div className={classes.Example}>
      <img
        src={`http://localhost:8080/images/${imagePath}`}
        alt={`example ${name} image`}
        onClick={() => history.push(`/examples/${id}`)}
      />
      <div className={classes.ExampleInfo}>
        <h3>{name}</h3>
        <p>{description}</p>
      </div>
    </div>
  );

  useEffect(() => {
    let isActive = true;
    const getExamples = async () => {
      const [data, newError] = await fetchExamples();

      if (!isActive) return;
      if (data) setExamples(data);
      setError(newError);
    };
    getExamples();

    return () => {
      isActive = false;
    };
  }, []);

  return (
    <div className={classes.index}>
      {error && <Popup text={TEXT_POPUP} onClick={() => history.push('/')} />}
      <LinkBack link="/" />
      <h1>{TEXT_H1}</h1>
      <div className={classes.ExamplesContainer}>
        {examples.map((e) => (
          <Example {...e} key={getId('example')} />
        ))}
      </div>
    </div>
  );
};

export default Library;
