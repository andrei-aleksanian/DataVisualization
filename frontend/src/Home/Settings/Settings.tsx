import Slider from 'Components/Forms/Slider';
import { useState } from 'react';
import { LinkBack } from 'Components/Link';

import { Algorithm } from 'types/Settings';
import CheckBoxes from 'Components/Forms/CheckBoxes';

import classes from './Settings.module.scss';
import COVA, { defaultSettingsCOVA } from './components/COVA';
import ANGEL, { defaultSettingsANGEL } from './components/ANGEL';

export const TEXT_CHECKBOX_COVA = 'COVA';
export const TEXT_CHECKBOX_ANGEL = 'ANGEL';
export const TEXT_H1 = 'Example: Cylinder';
export const TEXT_CHECKBOX_ALGORITHM = 'Algorithm:';
export const TEXT_SLIDER_NEIGHBOUR = 'Neighbour:';
export const TEXT_SLIDER_LAMBDA = 'Lambda:';
export const TEXT_CHECKBOX_PRESERVATION = 'Show Data Preservation:';

export enum DataPreservation {
  ON,
  OFF,
}
export interface SettingsCommon {
  dataPreservation: DataPreservation;
  algorithm: Algorithm;
  neighbour: number;
  lambda: number;
}
export const defaultSettingsCommon = {
  dataPreservation: DataPreservation.OFF,
  algorithm: Algorithm.COVA,
  neighbour: 0,
  lambda: 0,
};

const Settings = () => {
  const [settingsCommon, setSettingsCommon] = useState(defaultSettingsCommon);
  const [settingsCOVA, setSettingsCOVA] = useState(defaultSettingsCOVA);
  const [settingsANGEL, setSettingsANGEL] = useState(defaultSettingsANGEL);

  const onChangeAlgorithm = (event: React.ChangeEvent, newAlgorithm: Algorithm) => {
    event.preventDefault();
    setSettingsCommon((prev) => ({ ...prev, algorithm: newAlgorithm }));
  };
  const onChangeNeighbour = (value: number) => setSettingsCommon((prev) => ({ ...prev, neighbour: value }));
  const onChangeLambda = (value: number) => setSettingsCommon((prev) => ({ ...prev, lambda: value }));
  const onChangeDataPreservation = (event: React.ChangeEvent, value: DataPreservation) => {
    event.preventDefault();
    setSettingsCommon((prev) => ({ ...prev, dataPreservation: value }));
  };

  return (
    <div className={classes.index}>
      <LinkBack link="/" />
      <h1>{TEXT_H1}</h1>
      <CheckBoxes
        heading={TEXT_CHECKBOX_ALGORITHM}
        currentValue={settingsCommon.algorithm}
        onChange={onChangeAlgorithm}
        entries={[
          { value: Algorithm.COVA, text: TEXT_CHECKBOX_COVA },
          { value: Algorithm.ANGEL, text: TEXT_CHECKBOX_ANGEL },
        ]}
      />
      <Slider
        min={0}
        max={7}
        step={1}
        marksArr={['10', '20', '30', '10%', '20%', '30%', '40%', '50%']}
        onChange={onChangeNeighbour}
        text={TEXT_SLIDER_NEIGHBOUR}
        value={settingsCommon.neighbour}
      />
      <Slider
        min={0}
        max={1}
        step={0.2}
        marksArr={[0, 0.2, 0.4, 0.6, 0.8, 1]}
        onChange={onChangeLambda}
        text={TEXT_SLIDER_LAMBDA}
        value={settingsCommon.lambda}
      />
      {settingsCommon.algorithm === Algorithm.COVA ? (
        <COVA {...{ settingsCOVA, setSettingsCOVA }} />
      ) : (
        <ANGEL {...{ settingsANGEL, setSettingsANGEL }} />
      )}
      <CheckBoxes
        heading={TEXT_CHECKBOX_PRESERVATION}
        currentValue={settingsCommon.dataPreservation}
        onChange={onChangeDataPreservation}
        entries={[
          { value: DataPreservation.ON, text: 'Yes' },
          { value: DataPreservation.OFF, text: 'No' },
        ]}
      />
    </div>
  );
};

export default Settings;
