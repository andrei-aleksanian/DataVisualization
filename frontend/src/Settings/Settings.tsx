import Button from '../Components/Forms/Button';

import { Algorithm, SettingsInterface } from '../types/Settings';

import classes from './Settings.module.scss';

export const H1_TEXT = 'Demo Version';

export interface SettingsProps {
  setSettigns: React.Dispatch<React.SetStateAction<SettingsInterface>>;
  runAlgorithm: (e: React.MouseEvent) => Promise<void>;
  currentAlgorithm: Algorithm;
}

const Settings = ({ setSettigns, runAlgorithm, currentAlgorithm }: SettingsProps) => {
  const onClick = (event: React.MouseEvent, algorithm: Algorithm) => {
    event.preventDefault();
    setSettigns(() => ({
      algorithm,
    }));
  };

  return (
    <div className={classes.Settings}>
      <h1 className={classes.Heading}>{H1_TEXT}</h1>
      <Button
        text="COVA Dynamic"
        onClick={(e) => {
          onClick(e, Algorithm.COVA_PERSEVERANCE);
        }}
        active={currentAlgorithm === Algorithm.COVA_PERSEVERANCE}
        center
      />
      <Button
        text="Run the demo!"
        onClick={(e) => runAlgorithm(e)}
        customClass={classes.LongButton}
        long
        center
        active={false}
      />
    </div>
  );
};

export default Settings;
