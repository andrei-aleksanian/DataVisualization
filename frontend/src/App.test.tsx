import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';

import axios from 'utils/axios';
import { DATA_PERSEVERANCE } from 'types/Data/DataPerseverance';

import { placeholderExamples } from 'Home/Library/Library.test';
import App from './App';
import { TEXT_LINK_CUSTOM_DATA, TEXT_LINK_LIBRARY, TEXT_H1 as TEXT_H1_HOME } from './Home';
import { TEXT_H1 as TEXT_H1_EXAMPLES } from './Home/Settings';
import { TEXT_H1 as TEXT_H1_LIBRARY } from './Home/Library';

test('App matches snapshot', async () => {
  const { asFragment } = render(<App />, { wrapper: MemoryRouter });
  expect(asFragment()).toMatchSnapshot();
});

describe('Test navigation transitions', () => {
  const init = () => {
    render(<App />, { wrapper: MemoryRouter });
  };

  test('Custom data page transition', () => {
    init();
    const linkCustom = screen.getByText(TEXT_LINK_CUSTOM_DATA);
    userEvent.click(linkCustom);

    expect(screen.getByText('Sorry, this page is not implemented yet!')).toBeInTheDocument();
  });

  test('Library page transition', async () => {
    init();
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });
    const linkLibrary = screen.getByText(TEXT_LINK_LIBRARY);
    userEvent.click(linkLibrary);

    const libraryHeading = await waitFor(() => screen.getByText(TEXT_H1_LIBRARY));

    expect(libraryHeading).toBeInTheDocument();
  });

  test('Back to home from library', async () => {
    init();
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });
    const linkLibrary = screen.getByText(TEXT_LINK_LIBRARY);
    userEvent.click(linkLibrary);

    const linkBack = await waitFor(() => screen.getByAltText('back'));
    userEvent.click(linkBack);

    expect(screen.getByText(TEXT_H1_HOME)).toBeInTheDocument();
  });

  test('Examples page transition', async () => {
    init();
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });
    jest.spyOn(axios, 'post').mockResolvedValueOnce({ data: JSON.stringify(DATA_PERSEVERANCE) });

    const linkLibrary = await waitFor(() => screen.getByText(TEXT_LINK_LIBRARY));
    userEvent.click(linkLibrary);
    const linkExample = await waitFor(() => screen.getByAltText(`example ${placeholderExamples[0].name} image`));
    userEvent.click(linkExample);

    const heading = await waitFor(() => screen.getByText(TEXT_H1_EXAMPLES));
    expect(heading).toBeInTheDocument();
  });

  test('Back to library from examples', async () => {
    init();
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });

    const linkLibrary = screen.getByText(TEXT_LINK_LIBRARY);
    userEvent.click(linkLibrary);
    const linkExample = await waitFor(() => screen.getByAltText(`example ${placeholderExamples[0].name} image`));
    userEvent.click(linkExample);

    const linkBack = await waitFor(() => screen.getByAltText('back'));
    userEvent.click(linkBack);

    const libraryHeading = await waitFor(() => screen.getByText(TEXT_H1_LIBRARY));
    expect(libraryHeading).toBeInTheDocument();
  });

  test('Back to library from errored exaxmples page', async () => {
    init();
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });
    jest.spyOn(axios, 'post').mockRejectedValueOnce(new Error('Something went wrong'));
    jest.spyOn(axios, 'get').mockResolvedValueOnce({ data: placeholderExamples });

    const linkLibrary = screen.getByText(TEXT_LINK_LIBRARY);
    userEvent.click(linkLibrary);
    const linkExample = await waitFor(() => screen.getByAltText(`example ${placeholderExamples[0].name} image`));
    userEvent.click(linkExample);

    const linkBack = await waitFor(() => screen.getByText('OK'));
    userEvent.click(linkBack);

    const libraryHeading = await waitFor(() => screen.getByText(TEXT_H1_LIBRARY));
    expect(libraryHeading).toBeInTheDocument();
  });

  test('Back to home from custom data page', () => {
    init();
    const linkCustom = screen.getByText(TEXT_LINK_CUSTOM_DATA);
    userEvent.click(linkCustom);

    const linkBack = screen.getByAltText('back');
    userEvent.click(linkBack);

    expect(screen.getByText(TEXT_H1_HOME)).toBeInTheDocument();
  });
});
