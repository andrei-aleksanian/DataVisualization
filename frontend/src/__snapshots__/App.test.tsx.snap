// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="index"
    data-testid="Home"
  >
    <div
      class="Container"
    >
      <h1>
        COVA/ANGEL Visualization
      </h1>
      <p>
        Lorem ipsum dolor sit, amet consectetur adipisicing elit. Delectus dolorum laudantium nisi cupiditate pariatur odio deserunt sunt beatae nostrum tempore cumque iusto, quas ducimus vitae voluptatem repudiandae animi, repellendus commodi! Lorem ipsum dolor sit, amet consectetur adipisicing elit. Delectus dolorum laudantium nisi cupiditate pariatur odio deserunt sunt beatae nostrum tempore cumque iusto, quas ducimus vitae voluptatem repudiandae animi, repellendus commodi!
      </p>
      <a
        class="index Action"
        href="/examples"
      >
        See an Example
      </a>
      <a
        class="index Action"
        href="/your-data"
      >
        Visualize Your Data
      </a>
    </div>
  </div>
</DocumentFragment>
`;
